### DISCUSSION TITLE
initialData isn't always fresh for some reason

### DISCUSSION THREAD

**Discussion by dihmeetree** [trpc/trpc#5432] [General]
I'm using tRPC via Next.js (RSC's specifically)

I have an index route like so (`app/page.tsx`):
```ts
import { Suspense } from "react";

import { api } from "~/trpc/server";
import { CreatePostForm, PostList } from "./components/posts";
import Spinner from "./components/spinner";

async function Posts() {
  const posts = await api.post.all();
  return <PostList posts={posts} />;
}

export default async function HomePage() {
  return (
    <>
      <CreatePostForm />
      <Suspense fallback={<Spinner />}>
        <div className="w-full max-w-2xl overflow-y-scroll">
          <Posts />
        </div>
      </Suspense>
    </>
  );
}
```

And a PostList component like so:
```ts
export function PostList(props: { posts: RouterOutputs["post"]["all"] }) {
  const [posts] = api.post.all.useSuspenseQuery(undefined, {
    initialData: props.posts,
    refetchOnMount: false,
  });

  if (posts.length === 0) {
    return (
      <div className="relative flex w-full flex-col gap-4">
        <PostCardSkeleton pulse={false} />
        <PostCardSkeleton pulse={false} />
        <PostCardSkeleton pulse={false} />
        <div className="absolute inset-0 flex flex-col items-center justify-center bg-black/10">
          <p className="text-2xl font-bold text-white">No posts yet</p>
        </div>
      </div>
    );
  }

  return (
    <div className="flex w-full flex-col gap-4">
      {posts.map((p) => {
        return <PostCard key={p.id} post={p} />;
      })}
    </div>
  );
}
```

I'm getting fresh data from the RSC here (i.e - `await api.post.all()`). But for some reason when the fresh data is getting passed to the `PostList` component via `initialData`, the new data isn't being shown.

By default `StaleTime` is set to 0, so all data coming into `PostList` via `initialData` should be fresh and used to populate the query cache, no?

I want fresh data from the RSC to be used as the `initialData` in the `useSuspenseQuery`. This query should never run on mount.. data supplied to this component always comes from the RSC, so it never has to fetch the data again on the client (unless i'm invalidating it after a mutation).

Not sure what i'm doing wrong here.

**Note**: Data is fresh when I refresh the page, but when navigating between pages... data in `initialData` is cached.

**Comment by KATT** [comment]
Initial data is the initial data so I think it won't override anything, you might need a hydration boundary rather than passing the data like this

(This is more react query territory than trpc if you wanna search for it)

**Reply by dihmeetree** [reply]
Hi KATT thanks for the response. So the issue stems from the https://github.com/t3-oss/create-t3-turbo repo and i'm trying to resolve an issue where (similar to my original post), data is supplied via an RSC (i.e - `app/pag.tsx`) and then passed to a component via `initialData`, but then the `useQuery` fetches the same data again on the client side.

This seems like an unnecessary double fetch.. as fresh data is coming from the RSC, and doesn't need to fetch again client side. It seems that initialData is being cached somehow, so i'm trying to figure that out :)

**Reply by dihmeetree** [reply]
Nvm this is a issue with Next.js specifically... my issue is explained here https://medium.com/@CraftedX/why-i-hate-next-js-13-caching-4cd5a31d3a78

**Comment by dihmeetree** [comment]
@KATT does this look like a proper way to hydrate?

```ts
import { Suspense } from "react";
import { dehydrate, HydrationBoundary } from "@tanstack/react-query";
import { createServerSideHelpers } from "@trpc/react-query/server";
import superjson from "superjson";

import { appRouter } from "@acme/api";

import { createContext } from "~/trpc/server";
import { CreatePostForm, PostList } from "./components/posts";
import Spinner from "./components/spinner";

async function Posts() {
  const helpers = createServerSideHelpers({
    router: appRouter,
    ctx: await createContext(),
    transformer: superjson,
  });
  await helpers.post.all.prefetch();
  const dehydratedState = dehydrate(helpers.queryClient);

  return (
    <HydrationBoundary state={dehydratedState}>
      <PostList />
    </HydrationBoundary>
  );
}

export default async function HomePage() {
  return (
    <>
      <CreatePostForm />
      <Suspense fallback={<Spinner />}>
        <div className="w-full max-w-2xl overflow-y-scroll">
          <Posts />
        </div>
      </Suspense>
    </>
  );
}
```

It seems to now do what I'm looking for, so that's great.. just want to make sure this is the proper implementation.
